File: packs/dsa-trees-py/challenges/05-invert-binary-tree.test.py
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def build_tree(values):
    """Build binary tree from level-order list. None means no node."""
    if not values or values[0] is None:
        return None
    root = TreeNode(values[0])
    queue = deque([root])
    i = 1
    while queue and i < len(values):
        node = queue.popleft()
        if i < len(values) and values[i] is not None:
            node.left = TreeNode(values[i])
            queue.append(node.left)
        i += 1
        if i < len(values) and values[i] is not None:
            node.right = TreeNode(values[i])
            queue.append(node.right)
        i += 1
    return root


def tree_to_list(root):
    """Convert tree to level-order list (with None for missing nodes)."""
    if not root:
        return []
    result = []
    queue = deque([root])
    while queue:
        node = queue.popleft()
        if node:
            result.append(node.val)
            queue.append(node.left)
            queue.append(node.right)
        else:
            result.append(None)
    while result and result[-1] is None:
        result.pop()
    return result


from solution import invert_tree


def test_full_tree():
    root = build_tree([4, 2, 7, 1, 3, 6, 9])
    result = invert_tree(root)
    assert tree_to_list(result) == [4, 7, 2, 9, 6, 3, 1]


def test_small_tree():
    root = build_tree([2, 1, 3])
    result = invert_tree(root)
    assert tree_to_list(result) == [2, 3, 1]


def test_empty_tree():
    result = invert_tree(None)
    assert result is None


def test_single_node():
    root = build_tree([1])
    result = invert_tree(root)
    assert tree_to_list(result) == [1]


def test_left_only():
    root = build_tree([1, 2])
    result = invert_tree(root)
    assert tree_to_list(result) == [1, None, 2]

File: packs/dsa-trees-ts/challenges/05-invert-binary-tree.test.ts
import { join } from 'node:path';
import { writeChallengeToTmp, importModule } from '../../test-helpers.js';

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val = 0, left: TreeNode | null = null, right: TreeNode | null = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

function buildTree(values: (number | null)[]): TreeNode | null {
  if (!values.length || values[0] === null) return null;
  const root = new TreeNode(values[0]);
  const queue: TreeNode[] = [root];
  let i = 1;
  while (queue.length && i < values.length) {
    const node = queue.shift()!;
    if (i < values.length && values[i] !== null) {
      node.left = new TreeNode(values[i] as number);
      queue.push(node.left);
    }
    i++;
    if (i < values.length && values[i] !== null) {
      node.right = new TreeNode(values[i] as number);
      queue.push(node.right);
    }
    i++;
  }
  return root;
}

function treeToList(root: TreeNode | null): (number | null)[] {
  if (!root) return [];
  const result: (number | null)[] = [];
  const queue: (TreeNode | null)[] = [root];
  while (queue.length) {
    const node = queue.shift()!;
    if (node) {
      result.push(node.val);
      queue.push(node.left);
      queue.push(node.right);
    } else {
      result.push(null);
    }
  }
  while (result.length && result[result.length - 1] === null) result.pop();
  return result;
}

const CHALLENGE = join(import.meta.dirname, '05-invert-binary-tree.json');

describe('Invert Binary Tree', () => {
  let invertTree: (root: TreeNode | null) => TreeNode | null;
  let cleanup: () => void;

  beforeAll(async () => {
    const tmp = writeChallengeToTmp(CHALLENGE);
    cleanup = tmp.cleanup;
    const mod = await importModule<{ invertTree: typeof invertTree }>(
      join(tmp.tmpDir, 'solution.ts')
    );
    invertTree = mod.invertTree;
  });

  afterAll(() => cleanup());

  it('inverts a full tree', () => {
    const result = invertTree(buildTree([4, 2, 7, 1, 3, 6, 9]));
    expect(treeToList(result)).toEqual([4, 7, 2, 9, 6, 3, 1]);
  });

  it('inverts a small tree', () => {
    const result = invertTree(buildTree([2, 1, 3]));
    expect(treeToList(result)).toEqual([2, 3, 1]);
  });

  it('returns null for empty tree', () => {
    expect(invertTree(null)).toBeNull();
  });

  it('returns single node unchanged', () => {
    const result = invertTree(buildTree([1]));
    expect(treeToList(result)).toEqual([1]);
  });

  it('moves left-only child to right', () => {
    const result = invertTree(buildTree([1, 2]));
    expect(treeToList(result)).toEqual([1, null, 2]);
  });
});
